feat: assign custom roles

sean-brydon opened the pull request:
## What does this PR do?

This PR adds the ability for people to assign sub-team roles to members in the team when pbac is enabled.

## Visual Demo (For contributors especially)

#### Image Demo (if applicable):

<img width="872" height="532" alt="CleanShot 2025-07-21 at 13 51 31" src="https://github.com/user-attachments/assets/4aab1aa5-74fc-413c-8022-8e144b5e0469" />

<img width="1660" height="1365" alt="CleanShot 2025-07-21 at 13 51 39" src="https://github.com/user-attachments/assets/a39a922a-7eeb-4efa-9da3-1ef741c7bc8b" />

## How to test: 
Enable PBAC on a org by adding pbac to the team features for that specific team in DB. 
Manually refresh the app to invalidate team features cache.
Visit roles and permissions in settings sidebar fo teams. Create a role with any permissions - youll want to check the role CRUD here.
Goto the team members page - check filters work 
Edit a user - move them to the new role you created
impersonate
Ensure you can edit other users
Filter by new role

logout 
login to teampro or any other account with pbac disabled
Ensure filters and chaning role works as expected

sean-brydon committed:
feat: assign custom roles

sean-brydon committed:
enable new flow to work for orgs + teams

sean-brydon committed:
correctly display role on members list

sean-brydon committed:
fix factued values

sean-brydon committed:
fix filtering

sean-brydon committed:
fix duplicate roles

sean-brydon commented on packages/features/pbac/services/role-management.factory.ts:
```diff
@@ -45,24 +46,25 @@
+    // Check if role is one of the default role IDs (values, not keys)
+    const isDefaultRole = Object.values(DEFAULT_ROLE_IDS).includes(role as any);
 
     if (isDefaultRole) {
-      await this.roleService.assignRoleToMember(DEFAULT_ROLE_IDS[role], membershipId);
+      await this.roleService.assignRoleToMember(role as string, membershipId);
     } else {
-      const roleExists = await this.roleService.roleBelongsToTeam(role, organizationId);
+      const roleExists = await this.roleService.roleBelongsToTeam(role as string, organizationId);
       if (!roleExists) {
         throw new RoleManagementError(
           "You do not have access to this role",
           RoleManagementErrorCode.INVALID_ROLE
         );
       }
-      await this.roleService.assignRoleToMember(role, membershipId);
+      await this.roleService.assignRoleToMember(role as string, membershipId);
     }
```
dismiss

sean-brydon committed:
remove logs

eunjae-lee reviewed (commented):
I tested it and it's working well! 

I see some type errors and NITs from CodeRabbit.
I'll revisit this !

devin-ai-integration[bot] committed:
fix: resolve TypeScript errors in custom roles functionality

- Add missing onSettingsOpen prop to MemberInvitationModalWithoutMembers
- Update updateRoleInCache function to handle both MembershipRole enum and custom role strings
- Fix cache update logic to properly handle mixed role types without breaking type safety
- Ensure proper type handling for both traditional and custom roles in EditMemberSheet

Co-Authored-By: sean@cal.com <Sean@brydon.io>

devin-ai-integration[bot] committed:
fix: correct role validation logic in PBACRoleManager

- Fix assignRole method to properly check if role is a default MembershipRole enum key
- Use DEFAULT_ROLE_IDS mapping to get the correct role ID for default roles
- Resolves unit test failure in role-management.factory.test.ts

Co-Authored-By: sean@cal.com <Sean@brydon.io>

joeauyeung reviewed (dismissed):
Working as expected. I have a few small blocking comments that should be addressed in this PR rather than a follow up.

sean-brydon commented on packages/features/ee/teams/components/MemberChangeRoleModal.tsx:
```diff
@@ -45,10 +45,18 @@ export const updateRoleInCache = ({
-            ...member,
-            role: member.id === memberId ? role : member.role,
-          })),
+          members: page.members.map((member) => {
+            if (member.id === memberId) {
+              const isTraditionalRole = Object.values(MembershipRole).includes(role as MembershipRole);
+              return {
+                ...member,
+                role: isTraditionalRole ? (role as MembershipRole) : member.role,
+                customRoleId: isTraditionalRole ? null : (role as string),
+                customRole: isTraditionalRole ? null : member.customRole,
+              };
+            }
+            return member;
+          }),
```
Good spot ! Fixing these

sean-brydon committed:
fix role state optimistic cache

sean-brydon committed:
move to using repo

sean-brydon committed:
fix assingment

sean-brydon committed:
handle input for default roles

sean-brydon committed:
Merge branch 'main' into fix/assigning-child-team-custom-roles

sean-brydon committed:
add todo

sean-brydon committed:
Merge remote-tracking branch 'refs/remotes/origin/fix/assigning-child-team-custom-roles' into fix/assigning-child-team-custom-roles

sean-brydon committed:
Merge branch 'main' into fix/assigning-child-team-custom-roles
